File: IsometricRPG/IsometricRPG/Entities/Enemy.swift
import SpriteKit

final class Enemy: Entity {
    enum State {
        case idle
        case chasing
        case attacking
    }

    var state: State = .idle
    var lastAttackTime: TimeInterval = 0
    let experienceValue: Int = 25

    init(worldPosition: CGPoint) {
        let node = Enemy.createEnemyNode()
        super.init(worldPosition: worldPosition, health: Constants.enemyMaxHealth, node: node)
        node.zPosition = Constants.ZPosition.entity
        node.name = "enemy"
    }

    private static func createEnemyNode() -> SKNode {
        let container = SKNode()

        // Body
        let body = SKShapeNode(rectOf: Constants.enemySize, cornerRadius: 2)
        body.fillColor = SKColor(red: 0.8, green: 0.2, blue: 0.2, alpha: 1)
        body.strokeColor = SKColor(red: 0.6, green: 0.1, blue: 0.1, alpha: 1)
        body.lineWidth = 1.5
        container.addChild(body)

        // Eyes
        let eyeSize: CGFloat = 3
        let leftEye = SKShapeNode(circleOfRadius: eyeSize)
        leftEye.fillColor = .white
        leftEye.strokeColor = .clear
        leftEye.position = CGPoint(x: -4, y: 4)
        container.addChild(leftEye)

        let rightEye = SKShapeNode(circleOfRadius: eyeSize)
        rightEye.fillColor = .white
        rightEye.strokeColor = .clear
        rightEye.position = CGPoint(x: 4, y: 4)
        container.addChild(rightEye)

        // Shadow
        let shadow = SKShapeNode(ellipseOf: CGSize(width: 16, height: 7))
        shadow.fillColor = SKColor.black.withAlphaComponent(0.3)
        shadow.strokeColor = .clear
        shadow.position = CGPoint(x: 0, y: -Constants.enemySize.height / 2)
        shadow.zPosition = Constants.ZPosition.shadow - Constants.ZPosition.entity
        container.addChild(shadow)

        // Health bar background
        let hpBg = SKShapeNode(rectOf: CGSize(width: 22, height: 3))
        hpBg.fillColor = SKColor(red: 0.3, green: 0.3, blue: 0.3, alpha: 0.8)
        hpBg.strokeColor = .clear
        hpBg.position = CGPoint(x: 0, y: Constants.enemySize.height / 2 + 6)
        hpBg.name = "hpBg"
        container.addChild(hpBg)

        let hpBar = SKShapeNode(rectOf: CGSize(width: 20, height: 2))
        hpBar.fillColor = .green
        hpBar.strokeColor = .clear
        hpBar.position = CGPoint(x: 0, y: Constants.enemySize.height / 2 + 6)
        hpBar.name = "hpBar"
        container.addChild(hpBar)

        return container
    }

    func updateAI(playerPosition: CGPoint, currentTime: TimeInterval) {
        let dist = IsometricMath.distance(worldPosition, playerPosition)

        switch state {
        case .idle:
            if dist < Constants.enemyDetectionRange / Constants.tileWidth {
                state = .chasing
            }
        case .chasing:
            if dist < Constants.enemyAttackRange / Constants.tileWidth {
                state = .attacking
            } else if dist > Constants.enemyDetectionRange / Constants.tileWidth * 1.5 {
                state = .idle
            } else {
                moveToward(playerPosition)
            }
        case .attacking:
            if dist > Constants.enemyAttackRange / Constants.tileWidth * 1.2 {
                state = .chasing
            }
        }
    }

    private func moveToward(_ target: CGPoint) {
        let dir = IsometricMath.direction(from: worldPosition, to: target)
        let speed = Constants.enemySpeed / Constants.tileWidth * (1.0 / 60.0) // approx per-frame
        worldPosition = CGPoint(
            x: worldPosition.x + dir.x * speed,
            y: worldPosition.y + dir.y * speed
        )
    }

    func canAttack(currentTime: TimeInterval) -> Bool {
        return state == .attacking && (currentTime - lastAttackTime) >= Constants.enemyAttackCooldown
    }

    override func update(deltaTime: TimeInterval) {
        // Update health bar
        if let hpBar = node.childNode(withName: "hpBar") as? SKShapeNode {
            let ratio = CGFloat(health) / CGFloat(maxHealth)
            hpBar.xScale = max(ratio, 0)
            hpBar.fillColor = ratio > 0.5 ? .green : (ratio > 0.25 ? .yellow : .red)
        }

        node.zPosition = Constants.ZPosition.entity + (worldPosition.x + worldPosition.y) * 0.01

        super.update(deltaTime: deltaTime)
    }

    override func onDeath() {
        // Drop XP orb visual
        let orb = SKShapeNode(circleOfRadius: 5)
        orb.fillColor = .yellow
        orb.strokeColor = .orange
        orb.position = node.position
        orb.zPosition = Constants.ZPosition.entity
        orb.name = "xpOrb_\(experienceValue)"
        node.parent?.addChild(orb)

        let pulse = SKAction.sequence([
            SKAction.scale(to: 1.3, duration: 0.3),
            SKAction.scale(to: 0.8, duration: 0.3)
        ])
        let fadeAndRemove = SKAction.sequence([
            SKAction.wait(forDuration: 5.0),
            SKAction.fadeOut(withDuration: 0.5),
            SKAction.removeFromParent()
        ])
        orb.run(SKAction.group([SKAction.repeatForever(pulse), fadeAndRemove]))

        super.onDeath()
    }
}

File: IsometricRPGMono/Entities/Enemy.cs
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Graphics;

namespace IsometricRPG;

public enum EnemyState { Idle, Chasing, Attacking }

public class Enemy : Entity
{
    public EnemyState State          { get; private set; } = EnemyState.Idle;
    public double     LastAttackTime { get; set; }
    public int        XPReward       { get; set; } = 15;

    public Enemy(Vector2 worldPosition)
        : base(worldPosition, Constants.EnemyMaxHealth) { }

    /// Main AI tick. Call every frame with the current player world position, game time, and delta time.
    public void UpdateAI(Vector2 playerWorldPos, double currentTime, WorldManager world, double deltaTime)
    {
        float dist = IsometricMath.Distance(WorldPosition, playerWorldPos);

        // Convert pixel ranges to tile units
        float detectionTiles = Constants.EnemyDetectionRange / Constants.TileWidth;
        float attackTiles    = Constants.EnemyAttackRange    / Constants.TileWidth;

        if (dist > detectionTiles)
        {
            State = EnemyState.Idle;
            return;
        }

        if (dist <= attackTiles)
        {
            State = EnemyState.Attacking;
            return;
        }

        // Chasing
        State = EnemyState.Chasing;
        MoveToward(playerWorldPos, world, deltaTime);
    }

    private void MoveToward(Vector2 target, WorldManager world, double deltaTime)
    {
        var dir    = IsometricMath.Direction(WorldPosition, target);
        float step = (Constants.EnemySpeed / Constants.TileWidth) * (float)deltaTime;
        var desired = WorldPosition + dir * step;

        // Wall-slide
        if (world.IsWalkable(WorldPosition, desired))
        {
            WorldPosition = desired;
        }
        else
        {
            var slideX = new Vector2(desired.X, WorldPosition.Y);
            var slideY = new Vector2(WorldPosition.X, desired.Y);
            if (world.IsWalkable(WorldPosition, slideX))
                WorldPosition = slideX;
            else if (world.IsWalkable(WorldPosition, slideY))
                WorldPosition = slideY;
        }
    }

    public bool CanAttack(double currentTime)
        => State == EnemyState.Attacking
           && (currentTime - LastAttackTime) >= Constants.EnemyAttackCooldown;

    public override void Update(GameTime gameTime)
    {
        base.Update(gameTime);
    }

    public override void Draw(SpriteBatch spriteBatch, SpriteManager spriteManager, Vector2 cameraPos)
    {
        var screenPos = ScreenPosition - cameraPos;
        var size = Constants.EnemySize;

        // Body circle
        int diameter = (int)Math.Max(size.X, size.Y);
        var bodyColor = TintColor == Color.White ? new Color(180, 60, 60) : TintColor;  // dull red
        var circleTex = spriteManager.GetCircle(diameter, bodyColor);

        spriteBatch.Draw(
            circleTex,
            new Vector2(screenPos.X - diameter / 2f, screenPos.Y - diameter / 2f),
            Color.White);

        // Health bar
        DrawHealthBar(spriteBatch, spriteManager, screenPos, size);
    }

    private void DrawHealthBar(SpriteBatch sb, SpriteManager sm, Vector2 screenPos, Vector2 size)
    {
        const int barW = 22, barH = 3;
        float ratio = Math.Max(0f, (float)Health / MaxHealth);
        var bgTex   = sm.GetRect(barW, barH, new Color(60, 60, 60));
        var fgColor = ratio > 0.5f ? Color.Green : (ratio > 0.25f ? Color.Yellow : Color.Red);
        var fgTex   = sm.GetRect(Math.Max(1, (int)(barW * ratio)), barH, fgColor);

        var barOrigin = new Vector2(screenPos.X - barW / 2f, screenPos.Y - size.Y / 2f - 7f);
        sb.Draw(bgTex, barOrigin, Color.White);
        if (ratio > 0f)
            sb.Draw(fgTex, barOrigin, Color.White);
    }
}
